macros/src/main/scala/scala/tools/abide/traversal/WarningsMacro.scala

```diff
@@ -0,0 +1,73 @@
+package scala.tools.abide.traversal
+
+import scala.reflect.macros._
+import scala.language.experimental.macros
+
+/**
+ * WarningExtractor
+ *
+ * Provide the extraction macro that uses the placeholder [[WarningExtractor.tree]] value in the
+ * [[WarningExtractor]] class with to create a function from Tree to String by replacing the
+ * dummy tree with the function parameter.
+ *
+ * To get the String we use the StringContext that we assume is the macro caller and
+ * apply the `StringContext.s` method to the transformed arguments. We also assume all
+ * references to `_.this.tree` refer to the [[WarningExtractor.tree]] value as checking for
+ * enclosing context is non-trivial and discouraged.
+ */
+object WarningExtractor {
```

Comment:
I must admit it makes the whole warning definition kind of magical, but I feel it makes rules clearer and easier to read and write. I especially like being able to say `nok(tree)` instead of having to pass a `Warning` instance along that clobbers everything up.